File: 2024/src/day2/day2main.go
package day2

import (
	"regexp"
	"slices"
	"strconv"
)

/*
part1:
	318 - too high
	223 - too low
part2:
*/

func Day2Pt1(inputLines []string) int {
	// reports := parseInput("day2_test.txt")
	// reports := parseInput("day2.txt")
	reports := parseInput(inputLines)
	safeReportCount := 0
	for _, report := range reports {
		reportSafe := checkReportSafe(report)
		if reportSafe {
			safeReportCount++
		}
	}
	return safeReportCount
}

func Day2Pt2(inputLines []string) int {
	// reports := parseInput("day2_test.txt")
	// reports := parseInput("day2.txt")
	reports := parseInput(inputLines)
	safeReportCount := 0
	for i := 0; i < len(reports); i++ {
		reportSafe := checkReportSafe2(reports[i])
		if reportSafe {
			safeReportCount++
		}
	}
	return safeReportCount
}

func checkReportSafe2(report []int) bool {
	reportSafe := checkReportSafe(report)
	for i := 0; !reportSafe && i < len(report); i++ {
		/*
			Remove every level and retest
				(brute force)
		*/
		dampReport := slices.Delete(slices.Clone(report), i, i+1)
		reportSafe = checkReportSafe(dampReport)
		if reportSafe {
			break
		}
	}
	return reportSafe
}

func checkReportSafe(report []int) bool {
	diffs := []int{}
	prevLvl := report[0]
	for i := 1; i < len(report); i++ {
		currDiff := prevLvl - report[i]
		diffs = append(diffs, currDiff)
		prevLvl = report[i]
	}

	reportSafe := true
	for i := 0; i < len(diffs); i++ {
		currDiff := diffs[i]
		if currDiff == 0 {
			reportSafe = false
		} else if i > 0 {
			prevDiff := diffs[i-1]
			if currDiff > 0 && prevDiff < 0 {
				reportSafe = false // down to up
			} else if currDiff < 0 && prevDiff > 0 {
				reportSafe = false // up to down
			}
		}
		if currDiff < 0 {
			currDiff = -currDiff
		}
		if currDiff > 3 {
			reportSafe = false
		}
		if !reportSafe {
			break
		}
	}
	return reportSafe
}

func parseInput(inputLines []string) [][]int {
	levelRx := regexp.MustCompile(`\d+`)
	reports := [][]int{}
	for _, currLine := range inputLines {
		lvlStrs := levelRx.FindAllString(currLine, -1)
		currLvls := []int{}
		for _, lvlStr := range lvlStrs {
			lvl, err := strconv.Atoi(lvlStr)
			if err != nil {
				panic(err)
			}
			currLvls = append(currLvls, lvl)
		}
		reports = append(reports, currLvls)
	}
	return reports
}

File: 2024/src/day2/day2-main.lua

local strUtil = require("../util/str-util")

local printf = require("../util/printf")

local function parseInput(inputLines)
   local reports = {}
   for _, inputLine in ipairs(inputLines) do
    local report = {}
    for reportStr in string.gmatch(inputLine, "(%d+)") do
      table.insert(report, tonumber(reportStr))
    end
    table.insert(reports, report)
   end
   return reports
end

local function checkReport(report)
  local diffs = {}
  local prevLvl = report[1]
  for i=2, #report do
    local currDiff = report[i] - prevLvl
    table.insert(diffs, currDiff)
    prevLvl = report[i]
  end
  local prevDiff = nil
  for _, diff in ipairs(diffs) do
    local absDiff = math.abs(diff)
    if absDiff < 1 or absDiff > 3 then
      return false
    end
    if prevDiff == nil then
      prevDiff = diff
    else
      if prevDiff > 0 and diff < 0 or prevDiff < 0 and diff > 0 then
        return false
      end
    end
  end
  return true
  -- printf("%s\n", strUtil.join(diffs, " "))
end

--[[
306 - correct
]]
local function day2Pt1(inputLines)
  local reports = parseInput(inputLines)
  local safeReportCount = 0
  for _, report in ipairs(reports) do
    local reportSafe = checkReport(report)
    if reportSafe then
      safeReportCount = safeReportCount + 1
    end
    -- printf("%s | %s\n", strUtil.join(report, " "), reportSafe)
  end
  return safeReportCount
end

local day2MainModule = {
  day2Pt1 = day2Pt1,
}

return day2MainModule
